core-backend/common/src/lib.rs

```diff
@@ -136,34 +140,230 @@ pub trait Environment<E: Ext + IntoExtInfo + 'static>: Sized {
     fn into_gas_amount(self) -> GasAmount;
 }
 
-pub trait OnSuccessCode<T, E> {
-    fn on_success_code<F>(self, f: F) -> Result<i32, E>
-    where
-        F: FnMut(T) -> Result<(), E>;
+pub struct ExtErrorProcessor<'a, E: Ext, T = ()> {
+    ext: &'a mut ErrorSavingExt<E>,
+    success: Option<T>,
 }
 
-impl<T, E, E2> OnSuccessCode<T, E> for Result<T, E2> {
-    fn on_success_code<F>(self, mut f: F) -> Result<i32, E>
+impl<'a, E: Ext, T> ExtErrorProcessor<'a, E, T> {
+    pub fn new(ext: &'a mut ErrorSavingExt<E>) -> Self {
+        Self { ext, success: None }
+    }
+
+    pub fn with<U, F>(mut self, f: F) -> Result<Self, U>
     where
-        F: FnMut(T) -> Result<(), E>,
+        F: FnOnce(&mut ErrorSavingExt<E>) -> Result<T, U>,
+        U: CoreError,
     {
-        match self {
-            Ok(t) => f(t).map(|_| 0),
-            Err(_) => Ok(1),
-        }
+        match f(self.ext) {
+            Ok(t) => {
+                self.success = Some(t);
+            }
+            Err(err) => {
+                err.into_ext_error()?;
+            }
+        };
+        Ok(self)
+    }
+
+    pub fn on_success<U, F>(mut self, f: F) -> Result<Self, U>
+    where
+        F: FnOnce(T) -> Result<(), U>,
+    {
+        self.success.take().map(f).transpose()?;
+        Ok(self)
+    }
+
+    pub fn error_len(self) -> u32 {
+        self.ext
+            .err
+            .as_ref()
+            .and_then(|err| err.as_ext_error())
+            .map(|err| err.encoded_size() as u32)
+            .unwrap_or(0)
     }
 }
 
-pub trait IntoErrorCode {
-    fn into_error_code(self) -> i32;
+// TODO: use for issue #911
+pub struct ErrorSavingExt<T: Ext> {
+    pub inner: T,
+    /// Ext error, if available.
+    pub err: Option<ExtError>,
 }
 
-impl<E> IntoErrorCode for Result<(), E> {
-    fn into_error_code(self) -> i32 {
-        match self {
-            Ok(()) => 0,
-            // TODO: actual error codes
-            Err(_) => 1,
-        }
+impl<E> ErrorSavingExt<E>
+where
+    E: Ext,
+{
+    pub fn new(inner: E) -> Self {
+        Self { inner, err: None }
+    }
+
+    /// Return result and store error info in field
+    fn return_and_store_err<T>(&mut self, result: Result<T, E::Error>) -> Result<T, E::Error> {
+        result.map_err(|err| {
+            self.err = err.as_ext_error().cloned();
+            err
+        })
+    }
+}
+
+impl<T> Ext for ErrorSavingExt<T>
+where
+    T: Ext,
+{
+    type Error = T::Error;
+
+    fn alloc(
+        &mut self,
+        pages: WasmPageNumber,
+        mem: &mut dyn Memory,
+    ) -> Result<WasmPageNumber, Self::Error> {
+        let res = self.inner.alloc(pages, mem);
+        self.return_and_store_err(res)
+    }
+
+    fn block_height(&mut self) -> Result<u32, Self::Error> {
```

Comment:
Is there any way to make these wrappers general? Bad perspective to add a new wrapper per every new syscall.

Reply:
I would provide some trait to workaround different ext wrappers.
Something like

```rust
trait WithExt {
    with_ext<T, E>( .. ) -> Result<T, E>;
    with_ext_infallible<T>( .. ) -> T;
}
```

and impl for different essences:
```rust
impl WithExt for ErrorSavingExt {
    with_ext<T, E>( .. ) -> Result<T, E> {
        let res = self.ext.with( ... );
        self.return_and_store_error(res)
    }
}
```
etc

Does it seem like real solution and code base reducer?

Reply:
@ukint-vs @ark0f please, pay attention on that idea for your ext-summoning work